File: src/services/nativeGateway.ts
import {
  commands,
  type GatewayCatalogPreview,
  type GatewayLifecycleInput,
  type GatewayMutationResult,
  type GatewayImportPreview,
  type GatewayImportConfirmInput,
  type NativeModelSpec,
  type GatewayModelsSnapshot,
} from "../generated/bindings";
import { invokeGeneratedIpc, mapGeneratedCommandResponse } from "./generatedIpc";
import { toProviderSummary, validateProviderId } from "./providers/providers";
export type {
  GatewayCatalogPreview,
  GatewayLifecycleInput,
  GatewayMutationResult,
  GatewayImportPreview,
  GatewayImportConfirmInput,
  NativeModelSpec,
  GatewayModelsSnapshot,
};

function checkTarget<T extends { targetId: string }>(result: T, targetId: string): T {
  if (result.targetId !== targetId) throw new Error("IPC_NATIVE_TARGET_MISMATCH");
  return result;
}

function checkProvider(result: GatewayModelsSnapshot, providerId: number, providerUuid: string) {
  if (result.providerId !== providerId || result.providerUuid !== providerUuid || !result.revision)
    throw new Error("IPC_NATIVE_PROVIDER_MISMATCH");
  return result;
}

export function nativeGatewayModelsGet(providerId: number, providerUuid: string) {
  validateProviderId(providerId);
  return invokeGeneratedIpc<GatewayModelsSnapshot>({
    title: "读取网关模型声明失败",
    cmd: "native_gateway_models_get",
    args: { providerId },
    invoke: async () =>
      mapGeneratedCommandResponse(
        await commands.nativeGatewayModelsGet(providerId, providerUuid),
        (result) => checkProvider(result, providerId, providerUuid)
      ),
  });
}

export function nativeGatewayModelsSet(
  providerId: number,
  providerUuid: string,
  expectedRevision: string,
  models: NativeModelSpec[]
) {
  validateProviderId(providerId);
  return invokeGeneratedIpc<GatewayModelsSnapshot>({
    title: "保存网关模型声明失败",
    cmd: "native_gateway_models_set",
    args: { providerId },
    invoke: async () =>
      mapGeneratedCommandResponse(
        await commands.nativeGatewayModelsSet(providerId, providerUuid, expectedRevision, models),
        (result) => checkProvider(result, providerId, providerUuid)
      ),
  });
}

export function nativeGatewayCatalogPreview(targetId: string) {
  return invokeGeneratedIpc<GatewayCatalogPreview>({
    title: "读取 AIO 入口预览失败",
    cmd: "native_gateway_catalog_preview",
    args: { targetId },
    invoke: async () =>
      mapGeneratedCommandResponse(await commands.nativeGatewayCatalogPreview(targetId), (result) =>
        checkTarget(result, targetId)
      ),
  });
}

export function nativeGatewayApply(input: GatewayLifecycleInput) {
  return invokeGeneratedIpc<GatewayMutationResult>({
    title: "应用 AIO 入口失败",
    cmd: "native_gateway_apply",
    args: { targetId: input.targetId },
    invoke: async () =>
      mapGeneratedCommandResponse(await commands.nativeGatewayApply(input), (result) =>
        checkTarget(result, input.targetId)
      ),
  });
}

export function nativeGatewayRemove(input: GatewayLifecycleInput) {
  return invokeGeneratedIpc<GatewayMutationResult>({
    title: "移除 AIO 入口失败",
    cmd: "native_gateway_remove",
    args: { targetId: input.targetId },
    invoke: async () =>
      mapGeneratedCommandResponse(await commands.nativeGatewayRemove(input), (result) =>
        checkTarget(result, input.targetId)
      ),
  });
}

export function nativeGatewayImportPreview(targetId: string, nativeKey: string) {
  return invokeGeneratedIpc<GatewayImportPreview>({
    title: "预览原生网关导入失败",
    cmd: "native_gateway_import_preview",
    invoke: async () =>
      mapGeneratedCommandResponse(
        await commands.nativeGatewayImportPreview(targetId, nativeKey),
        (result) => {
          checkTarget(result, targetId);
          if (result.nativeKey !== nativeKey) throw new Error("IPC_NATIVE_PROVIDER_MISMATCH");
          return result;
        }
      ),
  });
}

export function nativeGatewayImportConfirm(input: GatewayImportConfirmInput) {
  // API keys and native source content must never enter IPC diagnostics.
  return invokeGeneratedIpc({
    title: "导入网关上游失败",
    cmd: "native_gateway_import_confirm",
    invoke: async () =>
      mapGeneratedCommandResponse(await commands.nativeGatewayImportConfirm(input), (rows) =>
        rows.map(toProviderSummary)
      ),
  });
}

File: src-tauri/src/commands/native_gateway.rs
//! Thin IPC boundary; generated nodes never receive frontend-supplied raw JSON.
use crate::app::native_gateway_service as service;
use crate::app_state::DbInitState;
use crate::domain::native_gateway::*;

#[tauri::command]
#[specta::specta]
pub(crate) async fn native_gateway_models_get(
    app: tauri::AppHandle,
    db_state: tauri::State<'_, DbInitState>,
    provider_id: i64,
    provider_uuid: String,
) -> Result<GatewayModelsSnapshot, String> {
    service::native_gateway_models_get(app, db_state, provider_id, provider_uuid).await
}
#[tauri::command]
#[specta::specta]
pub(crate) async fn native_gateway_models_set(
    app: tauri::AppHandle,
    db_state: tauri::State<'_, DbInitState>,
    provider_id: i64,
    provider_uuid: String,
    expected_revision: String,
    models: Vec<NativeModelSpec>,
) -> Result<GatewayModelsSnapshot, String> {
    service::native_gateway_models_set(
        app,
        db_state,
        provider_id,
        provider_uuid,
        expected_revision,
        models,
    )
    .await
}
#[tauri::command]
#[specta::specta]
pub(crate) async fn native_gateway_catalog_preview(
    app: tauri::AppHandle,
    db_state: tauri::State<'_, DbInitState>,
    target_id: String,
) -> Result<GatewayCatalogPreview, String> {
    service::native_gateway_catalog_preview(app, db_state, target_id).await
}
#[tauri::command]
#[specta::specta]
pub(crate) async fn native_gateway_apply(
    app: tauri::AppHandle,
    db_state: tauri::State<'_, DbInitState>,
    input: GatewayLifecycleInput,
) -> Result<GatewayMutationResult, String> {
    service::native_gateway_mutate(app, db_state, input, false).await
}
#[tauri::command]
#[specta::specta]
pub(crate) async fn native_gateway_remove(
    app: tauri::AppHandle,
    db_state: tauri::State<'_, DbInitState>,
    input: GatewayLifecycleInput,
) -> Result<GatewayMutationResult, String> {
    service::native_gateway_mutate(app, db_state, input, true).await
}
#[tauri::command]
#[specta::specta]
pub(crate) async fn native_gateway_import_preview(
    app: tauri::AppHandle,
    db_state: tauri::State<'_, DbInitState>,
    target_id: String,
    native_key: String,
) -> Result<GatewayImportPreview, String> {
    service::native_gateway_import_preview(app, db_state, target_id, native_key).await
}
#[tauri::command]
#[specta::specta]
pub(crate) async fn native_gateway_import_confirm(
    app: tauri::AppHandle,
    db_state: tauri::State<'_, DbInitState>,
    input: GatewayImportConfirmInput,
) -> Result<Vec<crate::providers::ProviderSummary>, String> {
    service::native_gateway_import_confirm(app, db_state, input).await
}
